SimExLite/PMICalculators/SimpleScatteringPMICalculator.py

```diff
@@ -0,0 +1,11147 @@
+
+        :returns: The snapshot data.
+        :rtype: dict
+
+        """
+
+        xsnp = dict()
+        xsnp["Z"] = numpy.loadtxt(
+            os.path.join(path_to_snapshot, "Z.dat")
+        )  # Atom numbers
+        xsnp["T"] = numpy.loadtxt(os.path.join(path_to_snapshot, "T.dat"))  # Atom type
+        xsnp["uid"] = numpy.loadtxt(
+            os.path.join(path_to_snapshot, "uid.dat")
+        )  # Unique atom ID.
+        xsnp["r"] = numpy.loadtxt(
+            os.path.join(path_to_snapshot, "r.dat")
+        )  # Cartesian coordinates.
+        xsnp["v"] = numpy.loadtxt(
+            os.path.join(path_to_snapshot, "v.dat")
+        )  # Cartesian velocities.
+        xsnp["m"] = numpy.loadtxt(os.path.join(path_to_snapshot, "m.dat"))  # Masses.
+        xsnp["q"] = numpy.loadtxt(os.path.join(path_to_snapshot, "q.dat"))  # Ion charge
+        xsnp["ff"] = numpy.loadtxt(
+            os.path.join(path_to_snapshot, "f0.dat")
+        )  # Form factors of each atom type.
+        xsnp["Q"] = numpy.loadtxt(
+            os.path.join(path_to_snapshot, "Q.dat")
+        )  # Wavenumber grid for form factors.
+
+        return xsnp
+
+    def f_load_sample(self, sample_path):
+        sample = dict()
+        xfp = h5py.File(sample_path, "r")
+        xxx = xfp.get("Z")
+        sample["Z"] = xxx[()]
+        xxx = xfp.get("r")
+        sample["r"] = xxx[()]
+        xfp.close()
+        sample["selZ"] = dict()
+        for sel_Z in numpy.unique(sample["Z"]):
+            sample["selZ"][sel_Z] = numpy.nonzero(sample["Z"] == sel_Z)
+        sample["N"] = len(sample["Z"])
+
+        self.g_s2e["sample"] = sample
+
+    def f_save_data(self, dset, data):
+        xfp = h5py.File(self.g_s2e["setup"]["pmi_out"], "a")
+        try:
+            xfp.create_group(os.path.dirname(dset))
+        except:
+            1
+
+        xfp[dset] = data
+        xfp.close()
+
+    def f_rotate_sample(self):
+        # Init quaternion for rotation.
+        self.g_s2e["sample"]["rot_quaternion"] = numpy.array([0, 0, 0, 0])
+
+        # Set to random if desired.
+        if self.g_s2e["random_rotation"] is True:
+            self.g_s2e["sample"]["rot_quaternion"] = numpy.random.rand(4)
+            self.g_s2e["sample"]["rotmat"] = numpy.zeros((9,))
+            s2e_gen_randrot_quat(
+                self.g_s2e["sample"]["rot_quaternion"], self.g_s2e["sample"]["rotmat"]
+            )
+            s2e_rand_orient(self.g_s2e["sample"]["r"], self.g_s2e["sample"]["rotmat"])
+
+        self.f_save_data(
+            "/data/angle", self.g_s2e["sample"]["rot_quaternion"].reshape((1, 4))
+        )
+
+    def f_system_setup(self):
+        self.g_s2e["sys"]["r"] = self.g_s2e["sample"]["r"].copy()
+        self.g_s2e["sys"]["q"] = numpy.zeros(self.g_s2e["sample"]["Z"].shape)
+        self.g_s2e["sys"]["NE"] = self.g_s2e["sample"]["Z"].copy()
+        self.g_s2e["sys"]["Z"] = self.g_s2e["sample"]["Z"]
+        self.g_s2e["sys"]["Nph"] = 1e99
```

Comment:
This is a lot, is it intended?

Reply:
Heritage. It will be overwritten by a more sensible number.